design/proposal-to-add-boot-parameters-for-metal3-schema.md

```diff
@@ -0,0 +1,135 @@
+<!--
+ This work is licensed under a Creative Commons Attribution 3.0
+ Unported License.
+
+ http://creativecommons.org/licenses/by/3.0/legalcode
+-->
+
+# Proposal to add BOOT parameter in the metal3 schema
+
+## Table of Contents
+
+<!--ts-->
+   * [Proposal to add BOOT parameter in the metal3 schema](#title)
+      * [Table of Contents](#table-of-contents)
+      * [Summary](#summary)
+      * [Motivation](#motivation)
+         * [Goals](#goals)
+      * [Proposal](#proposal)
+         * [Implementation Details/Notes/Constraints](#implementation-detailsnotesconstraints)
+         * [Risks and Mitigations](#risks-and-mitigations)
+      * [Design Details](#design-details)
+         * [Work Items](#work-items)
+         * [Dependencies](#dependencies)
+         * [Test Plan](#test-plan)
+      * [References](#references)
+<!--te-->
+
+## Summary
+
+Enhancing the `metal3` schema by adding `BOOT` parameters that we received 
+from ironic introspection which will be helpful in performance based use cases.
+
+## Motivation
+
+To make the existing `metal3` schema compatible and scalable for heavy computation 
+based use cases.  
+
+### Goals
+
+To compare the introspection data fetched from `Ironic-Inspector` through 
+`Gopher-cloud client` library with `metal3` schema and add `BOOT` parameters 
+in `metal3` schema.
+
+## Proposal
+
+The comparison of hardware introspection data coming from `Ironic-Inspector` 
+and existing `metal3` schema for configuration of `baremetalhost` concludes 
+that few hardware parameters are missing in `metal3` schema that are fetched 
+from `Ironic-Inspector`. We propose to add `BOOT` parameters in `metal3` schema for 
+scalability and performance enhancement.
+
+### Implementation Details/Notes/Constraints
+
+Link for Existing `Metal3` Specs
+
+Please refer `metal3` spec for bare-metal:
+https://github.com/metal3-io/baremetal-operator/blob/master/deploy/crds/metal3.io_baremetalhosts_crd.yaml
+
+The investigation of the existing schema indicates that BareMetal-Operator 
+gets BareMetal node introspection data using an API call to `Ironic-Inspector` 
+through `gopher-cloud client` library. 
+
+Having fetched BareMetal Node introspection data from Ironic-Inspector and
+compared it with BareMetalHost schema of `metal3` `(metal3.io_baremetalhosts_crd.yaml`), 
+we recommend following parameters need to be added in schema of `metal3`.
+1. Add enhancement in the existing `metal3` schema for the following parameters:
+
+    Boot Parameters:
+    
+    * current_boot_mode: `uefi`
```

Comment:
I don't really see the use of this. In future, the user will be able to select the boot mode they want in the CR (once metal3-io/baremetal-operator#302 or something like it merges). Knowing the boot mode that was used to boot the inspection image is not that interesting.
A list of the *available* boot modes could be interesting. I'm not sure we can get that from inspection though. In fact, it's arguably too late at that point.

Reply:
See, Usually as an Operator/admin, I would like to know existing boot mode before setting another boot mode. This is lot more useful for planning server boot management and it's PXE management accordingly. Because PXE settings are dependent on boot modes, so understanding existing boot mode will be useful for operator. Also it avoids resetting of same boot mode also.

Reply:
Inspection can't tell you the existing boot mode though. It can only tell you that it was booted once in that mode some time in the past. That's why I'm not sure it's that useful.

Reply:
By looking at existing boot mode,  it is helpful for user to avoid same boot mode again.

Reply:
I agree with Digambar, If user wants to change boot mode later after, It would be really helpful.

Reply:
> That makes it more important to add a field to set the desired boot mode, which this proposal doesn't cover.

Noting for the record that that discussion is already happening simultaneously in #63

Reply:
> It's unfortunate that there is a window of time between inspection and provisioning where the current state may be shown incorrectly. That's the case for all of the data we get through inspection, though, and any of the rest of the hardware details can change, too. That makes it more important to add a field to set the desired boot mode, which this proposal doesn't cover. Is there a reason for not including that field?

Hey, We are addressing desired_boot_mode functionality as part of another PR #63

Reply:
> > That makes it more important to add a field to set the desired boot mode, which this proposal doesn't cover.
> 
> Noting for the record that that discussion is already happening simultaneously in #63

Yes

Reply:
I'm just confused, because the obvious way to store the "current boot mode" is in the Provisioning data, at the time you boot it, based on the settings the user provided via #63 and that you actually used to boot it. Nobody has yet explained why it would be better to store a boot mode that is anything but current in the HardwareDetails and call it "current boot mode".